File: trunk/esper/src/main/java/net/esper/eql/agg/NumIntegerSumAggregator.java
package net.esper.eql.agg;

import net.esper.eql.agg.AggregationMethod;
import net.esper.eql.core.MethodResolutionService;

/**
 * Sum for any number value.
 */
public class NumIntegerSumAggregator implements AggregationMethod
{
    private int sum;
    private long numDataPoints;

    public void enter(Object object)
    {
        if (object == null)
        {
            return;
        }
        numDataPoints++;
        Number number = (Number) object;
        sum += number.intValue();
    }

    public void leave(Object object)
    {
        if (object == null)
        {
            return;
        }
        numDataPoints--;
        Number number = (Number) object;
        sum -= number.intValue();
    }

    public Object getValue()
    {
        if (numDataPoints == 0)
        {
            return null;
        }
        return sum;
    }

    public Class getValueType()
    {
        return Integer.class;
    }

    public AggregationMethod newAggregator(MethodResolutionService methodResolutionService)
    {
        return methodResolutionService.makeSumAggregator(null);
    }
}

File: NEsper/eql/agg/NumIntegerSumAggregator.cs
///////////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2007 Esper Team. All rights reserved.                                /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
///////////////////////////////////////////////////////////////////////////////////////

using System;
using net.esper.eql.agg;
using net.esper.eql.core;

namespace net.esper.eql.agg
{
	/// <summary>Sum for any number value.</summary>
	public class NumIntegerSumAggregator : AggregationMethod
	{
	    private int sum;
	    private long numDataPoints;

	    public void Enter(Object item)
	    {
	        if (item == null)
	        {
	            return;
	        }
	        numDataPoints++;
            int number = Convert.ToInt32(item);
            sum += number;
	    }

	    public void Leave(Object item)
	    {
	        if (item == null)
	        {
	            return;
	        }
	        numDataPoints--;
	        int number = Convert.ToInt32(item) ;
	        sum -= number;
	    }

	    public Object Value
	    {
            get
            {
                if (numDataPoints == 0)
                {
                    return null;
                }
                return sum;
            }
	    }

	    public Type ValueType
	    {
            get { return typeof(int?); }
	    }

	    public AggregationMethod NewAggregator(MethodResolutionService methodResolutionService)
	    {
	        return methodResolutionService.MakeSumAggregator(null);
	    }
	}
} // End of namespace
